### PR TITLE
Livescript support!

### PR DESCRIPTION
**Pull request by colinta**
I think this is good to go, @smanolloff do you want to take a look before we merge into master?  @pepkin88 and I whipped this up to address #23

### CODE REVIEW

**Comment by pepkin88** [comment]
What do you think if I implemented a "pushing behavior" for backslash strings under a flag?
What I'm imagining is an option in the settings, set to `false` by default, which, after enabling, would convert `"prop": 1` directly to `\prop : 1` instead of `'prop': 1`. Personally, I'd find it useful, it would fit stylistic preferences at my work. But I also don't see it being that critical and it might be considered not practical enough. All of the code would probably be contained in the `livescript` method.

What do you think?

**Comment by colinta** [comment]
That would be easy, too, to pass in as a "custom arg"

```json
"source.livescript": {
  "custom": ["livescript", true],
  "quotes": [["'", "\""], ["'''", "\"\"\""]]
},
```

```python
def livescript(self, region, push_backslash=False)
```

**Comment by pepkin88** [comment]
Yes, although, I believe, currently you can only pass there an array as a list of arguments, like this:
```json
"source.livescript": {
  "custom": ["livescript", [true]],
  "quotes": [["'", "\""], ["'''", "\"\"\""]]
},
```

**Comment by colinta** [comment]
Oh right, that was deliberate, in case we needed to add more features/flags to that list.

**Comment by pepkin88** [comment]
I changed it to named arguments, as to better expose them in the config file.
I'll post a patch soon.

**Comment by colinta** [comment]
I dig it!

File: change_quotes.py [hunk 1]
```diff
@@ -104,6 +105,23 @@ def run_each(self, edit, sel_region):
...
+        if self.custom:
+            fn_name = self.custom[0]
+            try:
+                fn = getattr(self, fn_name)
+            except AttributeError:
+                return
+
+            try:
+                fn_kwargs = self.custom[1]
+                if not isinstance(fn_kwargs, dict):
```
Comment (colinta) [inline change_quotes.py]: I would expect this to fail (and show an error message in the console) rather than simply not work.

**Comment by pepkin88** [comment]
Here's my take on this feature, do you like it?

File: change_quotes.py (hunk 1 above)
Comment (pepkin88) [inline change_quotes.py]: I was seeing guards for the existence of the method and for the existence of the second item in the array, so I assumed there should also be a guard for the dict type of this item. Most of the time, when I mess something in the settings, the default values are taking over an everything works.
But we could raise some exception there.

Comment (colinta) [inline change_quotes.py]: Oh that's a good point - yeah I think I'll remove the guard for fetching the custom method, or at least show an error message.  Fine for now, anyway.

File: change_quotes.py [hunk 2]
```diff
@@ -444,7 +464,7 @@ def replace_quotes(self, region, replacement):
         debug("Replace: %s with %s" % (self.view.substr(region), replacement))
         self.view.replace(self.edit, region, replacement)
 
-    def escape_unescape(self, region, quote, replacement):
+    def escape_unescape(self, region, quote, replacement, replacement_re = None):
```
Comment (smanolloff) [inline change_quotes.py]: Could you also update the docstring with some info about `replacement_re`?

Comment (smanolloff) [inline change_quotes.py]: Also, I see both `replacement` and `quote` can now be None, would love to also see that mentioned in the docstring :)

**Comment by smanolloff** [comment]
Seems like a nice approach for keeping main logic decoupled from special cases like this one.
Ideally, one would only need to add his own function and thus extend, instead of change, the plugin workflow (nothing is ever perfect, though haha ;) but at least it is an attempt to get as close as possible)

I left you a couple of comments regarding the docstring as I would really like to avoid outdated/misleading docstrings :)

File: change_quotes.py (hunk 2 above)
Comment (pepkin88) [inline change_quotes.py]: Sure thing.

File: change_quotes.py [hunk 3]
```diff
@@ -516,7 +516,7 @@ def livescript(self, region, backslash_push = False):
             inner = self.view.substr(inner_region)
             replacement = "'%s'" % (inner)
             self.replace_quotes(region, replacement)
-            self.escape_unescape(inner_region, None, r"\\", r"\\$")
+            self.escape_unescape(inner_region, None, r"\\", replacement_re = r"\\$")
```
Comment (pepkin88) [inline change_quotes.py]: In case there would be added `quote_re` or some other parameter in the future.

File: change_quotes.py [hunk 4]
```diff
@@ -468,21 +495,57 @@ def escape_unescape(self, region, quote, replacement):
...
 
+    def livescript(self, region, backslash_push = False):
+        first_3_chars = self.view.substr(sublime.Region(region.begin(), region.begin() + 3))
+        first_char = first_3_chars[0]
+        if first_char == '\\':
+            inner_region = sublime.Region(region.begin() + 1, region.end())
+            inner = self.view.substr(inner_region)
+            replacement = "'%s'" % (inner)
+            self.replace_quotes(region, replacement)
+            self.escape_unescape(inner_region, None, r"\\", replacement_re = r"\\$")
```
Comment (pepkin88) [inline change_quotes.py]: In case there would be added quote_re or some other parameter in the future.

File: change_quotes.py (hunk 3 above)
Comment (pepkin88) [inline change_quotes.py]: In case there would be added quote_re or some other parameter in the future.

**Comment by smanolloff** [comment]
hey, do we have any progress on this? 
I can't test the new functionality, so I have no idea how it performs.
I will, however, test the rest (e.g. old functionality) for regression :) I will update my plugin right now with this code to try it out for a while.

**Comment by pepkin88** [comment]
Yes, sorry for lack of updates. I wanted to make some changes, but I was busy lately.

I was planning on making better use of configuration options, namely having an impact on order of stages by changing order in the plugin's settings. While doing it, I was trying to make use of the existing code.
I did most of it, but was blocked by one of two bugs with (un)escaping I encountered.

I plan to crack this and post results later this week.

**Comment by pepkin88** [comment]
Sorry, I didn't make it on time, I'll try to push some commits tomorrow or in 2 days.

**Comment by smanolloff** [comment]
Just so that you know, I have been using this for more than a month now and have encountered no issues regarding the old functionality. It introduces no regression :)

**Comment by pepkin88** [comment]
I'm sorry for keeping you waiting. I guess my procrastination won't give me a chance to finish it in a reasonable time. So I won't be blocking this anymore. Although I will at least try to write about my ideas and thoughts.

My goal was to make better foundation for possible additional languages, so that it would be easier to add a custom behavior without any need for changing other parts of code. Also, like I wrote earlier, I was trying to better use the configuration options, to have more impact on the order of the modes in cycles. It also should be more intuitive, as the existence and the placement of `\\` (not just `custom: livescript`) would be indicative of enabling the feature. Sounds nice, but probably nobody would be using it anyway, so it was almost purely for the sake of perfectionism.

But I also found 2 bugs regarding escaping/unescaping, but they are not related to this issue (although the way of fixing that could have impact on the code from this pull request).
The first one was occurring, when `'''` was before `'` and the string was `'''a\'\'\'b'''`. The script was trying to escape `'`s first, but they are already escaped. So later it was trying to unescape `'''`s making it `'a'''b'`, which is incorrect. The solution is simple, change places of escape and unescape. I was planning here to split the escape_unescape function to two distinct functions; better separation of concerns in my opinion and the optional params wouldn't have to be optional anymore.
The second bug appears when the string `'''""""'''` is cycled to the `"""\"\"\"""""` with the last `"` hanging. It could be solved by replacing the regex from the end of the string, but I struggled to come up with an elegant solution.

So I had a block and didn't finish any of this. But I guess it's better to release working versions earlier, than to polish it more that it is needed.
So I give a green light on this PR and let's solve other issues when they really occur.

**Comment by smanolloff** [comment]
Hey @colinta no worries, I certainly did not mean to put any pressure, rather just a comment.

Regarding the two edge cases -- very strange, they seem to work fine for me. Are you certain that this happened within a context, where `'''` and `"""` are actually delimiters (python)? Because, If I try those examples in a python source code, I can't reproduce the issues you speak of. If I try them out in another context (with only `'` and `"` string delimiters), it *kinda* looks like the issue you speak of, but still its different (and ok, as far as I could see). Maybe it was something in your config file?

**Comment by colinta** [comment]
Thanks to everyone!  We can improve this as issues arise.

**Pull request merged**